File: 1-Introductory_Problems/17-Chessboard_and_Queens/solution.cpp
#include <iostream>
#include <vector>

bool	validPos(std::vector<std::string> &map, std::vector<int> &pos, int j)
{
	bool	result;

	result = true;
	for (int i = j - 1; i >= 0 && result; --i)
	{
		if (pos[j] == pos[i]
			|| pos[i] + (j - i) == pos[j]
			|| pos[i] - (j - i) == pos[j])
			result = false;
	}
	return (result && map[pos[j]][j] == '.');
}

int	calWays(std::vector<std::string> &map, std::vector<int> &pos, int j)
{
	int	res;

	if (j >= 8)
		return (1);
	res = 0;
	for (int i = 0; i < 8; ++i)
	{
		pos[j] = i;
		if (validPos(map, pos, j))
			res += calWays(map, pos, j + 1);
	}
	return (res);
}

int	main(void)
{
	std::vector<std::string>	map;
	std::vector<int>			pos;
	std::string					temp;

	for (int i = 0; i < 8; ++i)
	{
		std::cin >> temp;
		map.push_back(temp);
	}
	pos = std::vector<int>(8, 0);
	std::cout << calWays(map, pos, 0);
	return (0);
}

File: 1-Introductory_Problems/17-Chessboard_and_Queens/solution.hs
import Control.Monad (replicateM)
import Data.Sequence qualified as S
import Data.Foldable ( Foldable(toList) )

calWays :: [String] -> S.Seq Int -> Int -> Int
calWays m pos i =
  if i >= 8
    then 1
    else
      sum $ map f [0..7]
        where f j =
                let newPos = S.adjust' (const j) i pos
                    g (idx, p) = p /= j && p + (i - idx) /= j && p - (i - idx) /= j
                    validPos = m !! j !! i == '.' && all g (zip [0..] $ take i $ toList newPos) in
                  if validPos
                    then calWays m newPos $ i + 1
                    else 0

main :: IO ()
main = do
  m <- replicateM 8 getLine
  print $ calWays m (S.replicate 8 0) 0
